Add a readme to Beagle Web gRPC (#79)
ZupIT/beagle-helpers: web/beagle-grpc/README.md

# Beagle WEB gRPC

## What is gRPC?
----------------

> It is an open source framework to perform RPC (Remote call procedure) calls, through the HTTP2 protocol in a two-way streaming format, using Protobufs as interfaces between client and server.

Reviewer: ```suggestion
It is an open source framework to perform RPC (Remote Call Procedure) calls, through the HTTP2 protocol in a two-way streaming format, using Protobufs as interfaces between client and server.
```

> Protobufs, or Protocol Buffers, are a structured data serialization mechanism created by Google, which aim to keep contracts always valid and consistent regarding the implementations that consume a gRPC service.

Reviewer: ```suggestion
Protobufs, or Protocol Buffers, are a structured data serialization mechanism created by Google, which aims to keep contracts always valid and consistent, regarding the implementations that consumes a gRPC service.
```

2. Then import the gRPC Client:
   - **React:** ./src/beagle/beagle-service.ts
     ```typescript
     import { createBeagleUIService } from '@zup-it/beagle-react'
     import { usingBeagleGrpcClient } from 'beagle-web-grpc'

     export default createBeagleUIService({
       baseUrl: 'grpc://',
       components: {},
       fetchData: usingBeagleGrpcClient({ proxyAddress: 'http://localhost:8081' })
     })
     ```
   - **Angular:** ./src/app/beagle.module.ts
     ```typescript
     import { BeagleModule } from '@zup-it/beagle-angular'
     import { usingBeagleGrpcClient } from 'beagle-web-grpc'

     @BeagleModule({
       baseUrl: 'grpc://',
       module: {
         path: './beagle-components.module',
         name: 'BeagleComponentsModule',
       },
       components: {},
       fetchData: usingBeagleGrpcClient({ proxyAddress: 'http://localhost:8081' })
     })
     export class Beagle {}
     ```
3. Start your proxy running the following command on the root of the project: `beagle-web-grpc start-proxy` or `beagle-web-grpc spx` or `beagle-web-grpc start-proxy --mode your-mode`.
4. Use the Beagle Web as usually.

> > **Observations**

Reviewer: ```suggestion
> **Note:**
```

> Make sure that your BFF Application is running.

## CLI
------

> As gRPC is not compatible with the current structure of the Web, to support the technology, it is necessary to use a Proxy to establish the communication between the REST application and the gRPC Backend application, so we created a CLI (command-line interface), **beagle-web-grpc**, to facilitate the use of the proxy and provide all the necessary settings.

Reviewer: ```suggestion
As gRPC is not compatible with the current structure of the Web, to support the technology, it is necessary to use a proxy to establish the communication between the REST application and the gRPC backend application, so we created a CLI (command-line interface), **beagle-web-grpc**, to facilitate the use of the proxy and provide all the necessary settings.
```

#### Available commands for **beagle-web-grpc**

| Command       | Alias      | Options     | Description                                                                                       |
| ------------- | ---------- | ----------- | ------------------------------------------------------------------------------------------------- |

> | `init`        | `i`        |             | Creates the configuration file which define the settings to be used by the Beagle Web gRPC Proxy. |

Reviewer: ```suggestion
| `init`        | `i`        |             | Creates the configuration file which defines the settings to be used by the Beagle Web gRPC Proxy. |
```

#### Configuration file *./beagle-grpc.config.json*

After running `beagle-web-grpc init` the file *./beagle-grpc.config.json* will be created where the command was executed, generating a file with this content:

```json
{
  "configs": [
    {
      "mode": "development",
      "grpcBackendAddress": "localhost:50051",
      "tlsCertificatePath": "",
      "tlsKeyPath": "",
      "runProxyOnPort": 8081
    },
    {
      "mode": "production",
      "grpcBackendAddress": "https://my-grpc.backend.com/address",
      "tlsCertificatePath": "",
      "tlsKeyPath": "",
      "runProxyOnPort": 8081
    }
  ]
}
```

##### Attributes of the configuration file *./beagle-grpc.config.json*

| Attribute               | Description                                                                   | 
| ----------------------- | ----------------------------------------------------------------------------- |
| `mode`                  | Defines the mode name for the settings to be used when running the gRPC Proxy |

> | `grpcBackendAddress`    | Defines the address of the backend server that communicates using gRPC. Remembering that all requests that match this route will be redirected to the Proxy |

Reviewer: ```suggestion
| `grpcBackendAddress`    | Defines the address of the backend server that communicates using gRPC. Keep it mind that all requests that match this route will be redirected to the proxy |
```

> | `tlsCertificatePath`    | Set path to TLS Certificate location                                          |

Reviewer: ```suggestion
| `tlsCertificatePath`    | Sets the path to the TLS Certificate location                                          |
```

> | `tlsKeyPath`            | Set path to TLS Key location                                                  |

Reviewer: ```suggestion
| `tlsKeyPath`            | Sets the path to the TLS Key location                                                  |
```

| `runProxyOnPort`        | Defines the Port on which the gRPC Proxy will run                             |

## Client
----------

> The gRPC Client for Beagle Web is what redirects grpc requests to the proxy, the client is used through `BeagleService` in React and the `Beagle` module in Angular, using the `fetchData` property, which Beagle allows you to create custom HTTP requests, so the use goes like this: 

Reviewer: ```suggestion
The gRPC Client for Beagle Web redirects gRPC requests to the proxy. The client is used through `BeagleService` in React and the `Beagle` module in Angular, using the `fetchData` property. Beagle allows you to create custom HTTP requests, so the use goes like this: 
```

```typescript
import { usingBeagleGrpcClient } from 'beagle-web-grpc'

// { ... } - your module or service code
{
  // { ... } - other attributes
  fetchData: usingBeagleGrpcClient({ proxyAddress: 'http://localhost:8081' })
}
```

The gRPC client has the following initialization options (`BeagleGrpcClientOptions`):

| Attribute             | Required?       | Default Value   | Description                                                                           |
| --------------------- | --------------- | --------------- | ------------------------------------------------------------------------------------- |

> | `proxyAddress`        | Yes             |                 | Address where the Proxy will be running for the client to communicate with the proxy  |

Reviewer: ```suggestion
| `proxyAddress`        | Yes             |                 | Address where the proxy will be running for the client to communicate with the proxy  |
```
